Keypad/Keypad.js

```diff
@@ -0,0 +1,326 @@
+						</Cell>
+					);
+				})}
+			</Layout>
+		);
+	}
+});
+
+/**
+ * A Keypad component with an Input to display the outcome.
+ *
+ * @class Keypad
+ * @memberof agate/Keypad
+ * @extends agate/Keypad.KeypadBase
+ * @ui
+ * @public
+ */
+const KeypadExtended = hoc((config, Wrapped) => {
+	return class extends React.Component {
+		static displayName = 'KeypadExtended';
+
+		static propTypes = /** @lends agate/Keypad.Keypad.prototype */ {
+			/**
+			 * Applies a disabled style and the control becomes non-interactive.
+			 *
+			 * @type {Boolean}
+			 * @default false
+			 * @public
+			 */
+			disabled: PropTypes.bool,
+
+			/**
+			 * Called when a button is clicked. Includes the 'key' key in its event payload, updates the state and the input value accordingly.
+			 *
+			 * @type {Function}
+			 * @param {Object} event
+			 * @public
+			 */
+			handleInputValue: PropTypes.func,
+
+			/**
+			 * Called when the input value is changed.
+			 *
+			 * @type {Function}
+			 * @param {Object} event
+			 * @public
+			 */
+			onChange: PropTypes.func,
+
+			/**
+			 * The value of the input.
+			 *
+			 * @type {String}
+			 * @public
+			 */
+			value: PropTypes.string
+		};
+
+		constructor (props) {
+			super(props);
+
+			this.state = {
+				charIndex: 0,
+				keypadInput: ''
+			};
+		}
+
+		handleInputValue = (keyValue) => {
+			const {charIndex, keypadInput} = this.state;
+			let newKeypadInput = keypadInput;
+			let newCharIndex;
+
+			switch (keyValue) {
+				case 'arrowleftturn':
+				case 'Backspace':
+					newCharIndex = charIndex;
+					newKeypadInput = newKeypadInput.substring(0, charIndex - 1) + newKeypadInput.substring(charIndex, newKeypadInput.length);
+					newCharIndex = newCharIndex - 1;
+					break;
+
+				case 'ArrowLeft':
+					if (charIndex >= 0) {
+						newCharIndex = charIndex;
+						newCharIndex = newCharIndex - 1;
+					}
+					break;
+
+				case 'ArrowRight':
+					newCharIndex = charIndex;
+					newCharIndex = newCharIndex + 1;
+					break;
+
+				case 'Delete':
+					newCharIndex = charIndex;
+					newKeypadInput = newKeypadInput.substring(0, charIndex) + newKeypadInput.substring(charIndex + 1, newKeypadInput.length);
+					newCharIndex = newCharIndex - 1;
+					break;
+
+				case 'ArrowUp':
+				case 'ArrowDown':
+					// do nothing;
+					break;
+
+				case 'phone':
+					// method to call dialed number (keypadInput);
+					break;
+
+				default:
+					newCharIndex = charIndex;
+					newKeypadInput = newKeypadInput.substring(0, charIndex) + keyValue + newKeypadInput.substring(charIndex);
+					newCharIndex = newCharIndex + 1;
+					break;
+			}
+
+			if (keypadInput !== newKeypadInput) {
+				this.props.onChange({value: newKeypadInput});
+			}
+
+			if (keyValue !== 'phone') {
+				this.setState({
+					keypadInput: newKeypadInput,
+					charIndex: newCharIndex
+				});
+			} else {
+				this.setState({
+					charIndex: 0,
+					keypadInput: ''
+				});
+			}
```

Comment:
This could even be simpler with the following:

```javascript
case 'phone':
    newKeypadInput = '';
    newCharIndex = 0;

// and just this down here, no conditional:

this.setState({
	keypadInput: newKeypadInput,
	charIndex: newCharIndex
});
```

if the `if (keypadInput !== newKeypadInput) {` check is a problem, add the `phone` exception there and it will be cleaner.